### PR TITLE
70 poreikio gauti duomenis peržiūra

### LINKED ISSUE #70: Poreikio gauti duomenis peržiūra
- [ ] Testai
- [ ] Perkelti dizaino šablonus
- [ ] Įtraukti likusius elementus

Komentarai ir šaltiniai bus daromi atskirai.

Dalis darbo jau padaryta: https://github.com/atviriduomenys/katalogas/blob/main/vitrina/requests/views.py

### PR DESCRIPTION
**Pull request by adp-atea**

### CODE REVIEW

File: vitrina/requests/models.py [hunk 1]
```diff
@@ -45,6 +46,17 @@ class Meta:
...
+    def get_status_label(self):
+        if self.status == 'CREATED':
+            return _("Pateiktas")
+        elif self.status == 'REJECTED':
+            return _("Atmestas")
+        elif self.status == 'ALREADY_OPENED':
+            return _("Jau atvertas")
+        elif self.status == 'ANSWERED':
+            return _('Atsakytas')
+        return ''
```
Comment (sirex) [inline vitrina/requests/models.py]: Use choices: https://docs.djangoproject.com/en/3.2/ref/models/fields/#choices

File: vitrina/requests/views.py [hunk 2]
```diff
@@ -14,3 +17,47 @@ class RequestListView(ListView):
...
+        structure = []
+        for struct in structure_data:
+            data = struct.split(",")
+            if data:
+                structure.append({
+                    "data_title": data[0],
+                    "dictionary_title": data[1],
+                    "data_type": data[2],
+                    "data_notes": data[3],
+                })
```
Comment (sirex) [inline vitrina/requests/views.py]: Please move all data mangling to `services.py`.

File: vitrina/requests/views.py [hunk 3]
```diff
@@ -14,3 +17,47 @@ class RequestListView(ListView):
 class RequestDetailView(DetailView):
     model = Request
     template_name = 'vitrina/requests/detail.html'
+
+    def get_context_data(self, **kwargs):
+        context_data = super().get_context_data(**kwargs)
+        request = context_data.get('request')
```
Comment (sirex) [inline vitrina/requests/views.py]:
```suggestion
        request: Request = self.object
```
Please annotate data types, it took me some time to figure out, what exactly is `request`. Type annotations helps to better understand the code.

File: tests/test_request_detail.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            "data_type": 'type1',
+            "data_notes": 'notes1',
+        },
+        {
+            "data_title": 'data2',
+            "dictionary_title": 'dictionary2',
+            "data_type": 'type2',
+            "data_notes": 'notes2',
+        }
+    ]
```
Comment (sirex) [inline tests/test_request_detail.py]: Rename to `tests/requests/test_views.py`.

File: tests/test_request_detail.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+from vitrina.datasets.factories import DatasetFactory
+from vitrina.requests.factories import RequestFactory
+
+
+@pytest.mark.django_db
+def test_request_detail_view(app: DjangoTestApp):
+    dataset = DatasetFactory()
+    request = RequestFactory(dataset_id=dataset.pk, is_existing=True, status="REJECTED", purpose="science,product",
+                             changes="format", format="csv, json, rdf",
+                             structure_data="data1,dictionary1,type1,notes1;data2,dictionary2,type2,notes2")
```
Comment (sirex) [inline tests/test_request_detail.py]:
```suggestion
    request = RequestFactory(
        dataset=dataset,
        is_existing=True,
        status="REJECTED",
        purpose="science,product",
        changes="format",
        format="csv, json, rdf",
        structure_data=(
            "data1,dictionary1,type1,notes1;"
            "data2,dictionary2,type2,notes2"
        )
    )
```

> Beautiful is better than ugly.
> Readability counts.
> -- https://peps.python.org/pep-0020/

File: vitrina/requests/views.py [hunk 6]
```diff
@@ -14,3 +17,47 @@ class RequestListView(ListView):
...
+            try:
+                dataset = Dataset.public.get(pk=request.dataset_id)
+            except ObjectDoesNotExist:
+                pass
+
+        liked = False
+        if self.request.user and self.request.user.pk:
+            user_like = UserLike.objects.filter(request_id=request.pk, user_id=self.request.user.pk)
+            if user_like.exists():
+                liked = True
```
Comment (sirex) [inline vitrina/requests/views.py]: Please move all data mangling to `services.py`.

> Layered designs in information systems are another embodiment of separation of concerns (e.g., presentation layer, business logic layer, data access layer, persistence layer)
> -- https://en.wikipedia.org/wiki/Separation_of_concerns

File: vitrina/requests/views.py [hunk 7]
```diff
@@ -14,3 +17,47 @@ class RequestListView(ListView):
...
+        dataset = None
+        if request.dataset_id:
+            try:
+                dataset = Dataset.public.get(pk=request.dataset_id)
+            except ObjectDoesNotExist:
+                pass
+
+        liked = False
+        if self.request.user and self.request.user.pk:
+            user_like = UserLike.objects.filter(request_id=request.pk, user_id=self.request.user.pk)
```
Comment (sirex) [inline vitrina/requests/views.py]:
```suggestion
            user_like = UserLike.objects.filter(request=request, user=self.request.user)
```

File: vitrina/requests/views.py [hunk 8]
```diff
@@ -14,3 +17,47 @@ class RequestListView(ListView):
...
+
+        dataset = None
+        if request.dataset_id:
+            try:
+                dataset = Dataset.public.get(pk=request.dataset_id)
+            except ObjectDoesNotExist:
+                pass
+
+        liked = False
+        if self.request.user and self.request.user.pk:
```
Comment (sirex) [inline vitrina/requests/views.py]:
```suggestion
        if self.rrequest.user.is_authenticated:
```

https://docs.djangoproject.com/en/3.2/topics/auth/default/#authentication-in-web-requests

**Review by sirex — CHANGES_REQUESTED** [review]

File: vitrina/requests/views.py (hunk 7 above)
Comment (adp-atea) [reply vitrina/requests/views.py]: UserLike doesn't have fields request and user, would need to change model.

Comment (sirex) [reply vitrina/requests/views.py]: Ok, lets leave it like that, user likes should be converted to generic model anyway.

File: vitrina/requests/services.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+            structure.append({
+                "data_title": data[0],
+                "dictionary_title": data[1],
+                "data_type": data[2],
+                "data_notes": data[3],
+            })
+    return structure
+
+
+def get_is_liked(user, request):
```
Comment (sirex) [inline vitrina/requests/services.py]: Please, use type annotations, it improves code readability and reliability a lot.

```suggestion
def get_is_liked(user: User, request: Request) -> bool:
```

File: vitrina/requests/services.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+from vitrina.likes.models import UserLike
+
+
+def get_structure(request):
```
Comment (sirex) [inline vitrina/requests/services.py]:
```suggestion
def get_structure(request: Request) -> List[Dict[str, Any]]:
```

File: vitrina/requests/templates/vitrina/requests/requestCoretable.html [hunk 11]
```diff
@@ -5,68 +5,55 @@
         <th>{% translate "Pasiūlymas organizacijai" %}</th>
         <td>{% if organization.slug %}<a href="/organizations/{{ organization.slug }}">{{ organization }}</a>{% else %}{{ organization }}{% endif %}</td>
     </tr>
-    {% if request.isExisting == 'true' %}
+    {% if request.is_existing and dataset %}
         <tr>
-            <th>{% translate "request.forDataset" %}</th>
-            <td>{% if datasetSlug %}<a href="/dataset/{{ datasetSlug }}">{{ request.title }}</a>{% else %}{{ request.title }}{% endif %}</td>
+            <th>{% translate "Norimas keisti duomenų rinkinys" %}</th>
+            <td>{% if dataset.slug %}<a href="/dataset/{{ dataset.slug }}">{{ request.title }}</a>{% else %}{{ request.title }}{% endif %}</td>
```
Comment (sirex) [inline vitrina/requests/templates/vitrina/requests/requestCoretable.html]:
```suggestion
            <td>{% if dataset.slug %}<a href="{{ dataset.get_absolute_url }}">{{ request.title }}</a>{% else %}{{ request.title }}{% endif %}</td>
```

File: vitrina/requests/templates/vitrina/requests/requestCoretable.html [hunk 12]
```diff
@@ -5,68 +5,55 @@
         <th>{% translate "Pasiūlymas organizacijai" %}</th>
         <td>{% if organization.slug %}<a href="/organizations/{{ organization.slug }}">{{ organization }}</a>{% else %}{{ organization }}{% endif %}</td>
```
Comment (sirex) [inline vitrina/requests/templates/vitrina/requests/requestCoretable.html]:
```suggestion
        <td>{% if organization.slug %}<a href="{{ organization.get_absolute_url }}">{{ organization }}</a>{% else %}{{ organization }}{% endif %}</td>
```

File: vitrina/requests/templates/vitrina/requests/requestCoretable.html [hunk 13]
```diff
@@ -5,68 +5,55 @@
...
-                  {#  {% include "../../component/like.twig" %}#}
-                {#{% endif %}#}
-            {#{% endif %}#}
+    <tr>
+        <th>{% translate "Patinka" %}:</th>
+        <td>
+            {{ like_count }}
+            {% if user.is_authenticated %}
+                {% if not liked %}
+                    <form action="/request/{{ request.pk }}/like" method="post" class="rating-form">
```
Comment (sirex) [inline vitrina/requests/templates/vitrina/requests/requestCoretable.html]: Please, reverse URL by URL name: https://docs.djangoproject.com/en/3.2/topics/http/urls/#reverse-resolution-of-urls

**Review by sirex — CHANGES_REQUESTED** [review]

**Comment by adp-atea** [comment]
Could you approve this?

**Review by sirex — APPROVED** [review]

**Pull request merged**